internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+const (
+	requestStart = `
+{
+  "streams": [
+    {
+      "stream": {
+        "label": "pebble"
+      },
+      "values": [`
+
+	requestEnd = `
+      ]
+    }
+  ]
+}`
+)
+
+// lokiClient implements logClient
+// We send a new request to Loki when either:
+//   - we have maxMessages ready to send; or
+//   - delay has passed since the last request.
+type lokiClient struct {
+	lokiURL     string
+	maxMessages int
+	delay       time.Duration
+
+	// buffered request so far
+	buffer *bytes.Buffer
+	// how many messages have been written so far
+	messages int
+	// notify when buffer fills up
+	writeCh chan struct{}
+	// notify when deadline elapses
+	timeCh <-chan time.Time
+}
+
+func newLokiClient(lokiURL string, maxMessages int, delay time.Duration) *lokiClient {
+	c := &lokiClient{
+		lokiURL:     lokiURL,
+		maxMessages: maxMessages,
+		delay:       delay,
+	}
+	c.Start()
+	return c
+}
+
+func (c *lokiClient) Start() {
+	c.reset()
+	go c.run()
+}
+
+func (c *lokiClient) reset() {
+	c.buffer = bytes.NewBuffer([]byte(requestStart))
+	c.messages = 0
+	c.writeCh = make(chan struct{})
+	c.timeCh = time.After(c.delay)
+}
+
+// Main loop
+func (c *lokiClient) run() {
+	for {
+		select {
+		case <-c.writeCh:
+			c.flush()
+		case <-c.timeCh:
+			c.flush()
+		}
+	}
+}
+
+// send currently written request to the remote Loki
+// and reset state
+func (c *lokiClient) flush() {
+	if c.messages == 0 {
+		return
+	}
+
+	c.buffer.WriteString(requestEnd)
+	logger.Noticef(c.buffer.String())
+	resp, err := http.Post(c.lokiURL, "application/json", c.buffer)
```

Comment:
We shouldn't use the default go http client. One should be made with the lokiClient.

Reply:
Why, what's wrong with the default client?
@hpidcock thoughts?

Reply:
It may or may not matter here, but I recommend not using `http.DefaultClient`, because you don't get control over timeouts (unless you set the global timeout) and connection pooling, and it's harder to test because you can't pass in a testing client. Similar to the problems with globals in general.

You can work around the timeout issue by using `http.NewRequestWithContext` and supplying a context without a timeout, though. Still, I'd advise against it, and prefer passing in the `*http.Client` "from above" as a dependency.

Reply:
ok sure, easy enough to use a custom client